### ISSUE TITLE
[SE-0347] Type inference for generic default expressions doesn't work across modules

### ISSUE THREAD

**Issue by ffried** [apple/swift#60575]
**Describe the bug**
The feature described in [SE-0347](https://github.com/apple/swift-evolution/blob/main/proposals/0347-type-inference-from-default-exprs.md) only works within a single module.
As soon as a function using this is used in another module, the compiler is not able to determine the type of the generic argument.

Simplified example:

**Module1**

```swift
public enum Object1 {
    public static func test(_: (some Hashable)? = Optional<Never>.none) {}
}
```

**Module2**
```swift
import Module1

func test() {
    Object1.test()
}
```

**Steps To Reproduce**
1. Build the attached sample package (either using Xcode or Swift build)

**Expected behavior**
The package builds without errors.

**Screenshots**
<img width="1605" alt="image" src="https://user-images.githubusercontent.com/1689782/184915168-d3b1a6a0-83ea-4eec-8aa9-3f1b2824dff0.png">

**Environment (please fill out the following information)**
 - OS: macOS 12.5
 - Xcode Version/Tag/Branch: Xcode 14 Beta 5

**Additional context**

[BrokenGenericDefaults.zip](https://github.com/apple/swift/files/9352313/BrokenGenericDefaults.zip)

**Comment by xedin** [comment]
Should be fixed by https://github.com/apple/swift/pull/60132. Could you please verify using the latest main or 5.7 compiler snapshot from swift.org?

**Comment by ffried** [comment]
@xedin You're right, the attached sample package builds using the 2022-08-02 snapshot for Swift 5.7 from swift.org.
I can't test the main project where I found the issue, though, because the snapshot doesn't seem to contain the new regex features:
<img width="727" alt="image" src="https://user-images.githubusercontent.com/1689782/185058921-c4a1d0ea-fe58-44c7-b9c0-876fb8f37aa3.png">


But either way, I'd consider this one fixed. Hopefully the next Xcode beta will contain it this fix?

**Comment by xedin** [comment]
Sorry but we cannot promise anything about betas but the fix has been merged into 5.7 branch.
